### PR TITLE
Specify backing image format

### PR DESCRIPTION
**Pull request by tacerus**
On new versions of QEMU, a "convert" call without this patch will fail with "Backing file specified without backing format":

Before:

```
2023-07-06 13:02:26,563: vrnetlab   DEBUG    Creating overlay disk image
qemu-img: /ffp-12.1X47-D15.4-packetmode-0-overlay.qcow2: Backing file specified without backing format
Detected format of qcow2.2023-07-06 13:02:26,587: vrnetlab   DEBUG    Starting vrnetlab <__main__.VSRX object at 0x7facd1bede50>
2023-07-06 13:02:26,587: vrnetlab   DEBUG    VMs: [<__main__.VSRX_vm object at 0x7facd1bee6d0>]
2023-07-06 13:02:26,589: vrnetlab   DEBUG    VM not started; starting!
2023-07-06 13:02:26,589: vrnetlab   INFO     Starting VSRX_vm[0]
2023-07-06 13:02:26,589: vrnetlab   DEBUG    ['qemu-system-x86_64', '-enable-kvm', '-display', 'none', '-machine', 'pc', '-monitor', 'tcp:0.0.0.0:4000,server,nowait', '-m', '2048', '
-serial', 'telnet:0.0.0.0:5000,server,nowait', '-drive', 'if=ide,file=/ffp-12.1X47-D15.4-packetmode-0-overlay.qcow2', '-smp', '2', '-device', 'pci-bridge,chassis_nr=1,id=pci.1', '-de
vice', 'virtio-net-pci,netdev=p00,mac=52:54:00:79:ab:00', '-netdev', 'user,id=p00,net=10.0.0.0/24,tftp=/tftpboot,hostfwd=tcp::2830-10.0.0.15:830,hostfwd=tcp::2080-10.0.0.15:80,hostfw
d=udp::2161-10.0.0.15:161,hostfwd=tcp::2443-10.0.0.15:443,hostfwd=tcp::2022-10.0.0.15:22', '-device', 'virtio-net-pci,netdev=p01,mac=52:54:00:a1:01:01,bus=pci.1,addr=0x2', '-netdev',
 'socket,id=p01,listen=:10001', '-device', 'virtio-net-pci,netdev=p02,mac=52:54:00:33:ff:02,bus=pci.1,addr=0x3', '-netdev', 'socket,id=p02,listen=:10002', '-device', 'virtio-net-pci,
netdev=p03,mac=52:54:00:8f:21:03,bus=pci.1,addr=0x4', '-netdev', 'socket,id=p03,listen=:10003', '-device', 'virtio-net-pci,netdev=p04,mac=52:54:00:34:53:04,bus=pci.1,addr=0x5', '-net
dev', 'socket,id=p04,listen=:10004', '-device', 'virtio-net-pci,netdev=p05,mac=52:54:00:25:da:05,bus=pci.1,addr=0x6', '-netdev', 'socket,id=p05,listen=:10005', '-device', 'virtio-net
-pci,netdev=p06,mac=52:54:00:07:7a:06,bus=pci.1,addr=0x7', '-netdev', 'socket,id=p06,listen=:10006', '-device', 'virtio-net-pci,netdev=p07,mac=52:54:00:64:3a:07,bus=pci.1,addr=0x8',
'-netdev', 'socket,id=p07,listen=:10007', '-device', 'virtio-net-pci,netdev=p08,mac=52:54:00:2d:a0:08,bus=pci.1,addr=0x9', '-netdev', 'socket,id=p08,listen=:10008', '-device', 'virti
o-net-pci,netdev=p09,mac=52:54:00:fa:d4:09,bus=pci.1,addr=0xa', '-netdev', 'socket,id=p09,listen=:10009', '-device', 'virtio-net-pci,netdev=p10,mac=52:54:00:0c:9a:0a,bus=pci.1,addr=0
xb', '-netdev', 'socket,id=p10,listen=:10010']
2023-07-06 13:02:26,639: vrnetlab   INFO     STDOUT:
2023-07-06 13:02:26,639: vrnetlab   INFO     STDERR: qemu-system-x86_64: -drive if=ide,file=/ffp-12.1X47-D15.4-packetmode-0-overlay.qcow2: Could not open '/ffp-12.1X47-D15.4-packetmo
de-0-overlay.qcow2': No such file or directory
```

After:

```
2023-07-06 13:19:14,964: vrnetlab   DEBUG    Creating overlay disk image
2023-07-06 13:19:14,994: vrnetlab   DEBUG    Starting vrnetlab <__main__.VSRX object at 0x7f21da75e4d0>
2023-07-06 13:19:14,994: vrnetlab   DEBUG    VMs: [<__main__.VSRX_vm object at 0x7f21da75e550>]
2023-07-06 13:19:14,996: vrnetlab   DEBUG    VM not started; starting!
2023-07-06 13:19:14,996: vrnetlab   INFO     Starting VSRX_vm[0]
2023-07-06 13:19:14,996: vrnetlab   DEBUG    ['qemu-system-x86_64', '-enable-kvm', '-display', 'none', '-machine', 'pc', '-monitor', 'tcp:0.0.0.0:4000,server,nowait', '-m', '2048', '
-serial', 'telnet:0.0.0.0:5000,server,nowait', '-drive', 'if=ide,file=/ffp-12.1X47-D15.4-packetmode-0-overlay.qcow2', '-smp', '2', '-device', 'pci-bridge,chassis_nr=1,id=pci.1', '-de
vice', 'virtio-net-pci,netdev=p00,mac=52:54:00:fb:aa:00', '-netdev', 'user,id=p00,net=10.0.0.0/24,tftp=/tftpboot,hostfwd=tcp::2443-10.0.0.15:443,hostfwd=tcp::2022-10.0.0.15:22,hostfw
d=udp::2161-10.0.0.15:161,hostfwd=tcp::2830-10.0.0.15:830,hostfwd=tcp::2080-10.0.0.15:80', '-device', 'virtio-net-pci,netdev=p01,mac=52:54:00:39:1b:01,bus=pci.1,addr=0x2', '-netdev',
 'socket,id=p01,listen=:10001', '-device', 'virtio-net-pci,netdev=p02,mac=52:54:00:1c:b9:02,bus=pci.1,addr=0x3', '-netdev', 'socket,id=p02,listen=:10002', '-device', 'virtio-net-pci,
netdev=p03,mac=52:54:00:7f:2e:03,bus=pci.1,addr=0x4', '-netdev', 'socket,id=p03,listen=:10003', '-device', 'virtio-net-pci,netdev=p04,mac=52:54:00:d2:f7:04,bus=pci.1,addr=0x5', '-net
dev', 'socket,id=p04,listen=:10004', '-device', 'virtio-net-pci,netdev=p05,mac=52:54:00:25:cc:05,bus=pci.1,addr=0x6', '-netdev', 'socket,id=p05,listen=:10005', '-device', 'virtio-net
-pci,netdev=p06,mac=52:54:00:12:33:06,bus=pci.1,addr=0x7', '-netdev', 'socket,id=p06,listen=:10006', '-device', 'virtio-net-pci,netdev=p07,mac=52:54:00:19:a2:07,bus=pci.1,addr=0x8',
'-netdev', 'socket,id=p07,listen=:10007', '-device', 'virtio-net-pci,netdev=p08,mac=52:54:00:34:c4:08,bus=pci.1,addr=0x9', '-netdev', 'socket,id=p08,listen=:10008', '-device', 'virti
o-net-pci,netdev=p09,mac=52:54:00:1b:92:09,bus=pci.1,addr=0xa', '-netdev', 'socket,id=p09,listen=:10009', '-device', 'virtio-net-pci,netdev=p10,mac=52:54:00:ed:c3:0a,bus=pci.1,addr=0
xb', '-netdev', 'socket,id=p10,listen=:10010']
2023-07-06 13:19:50,511: launch     INFO     VM started
```

### CODE REVIEW

**Comment by mzagozen** [comment]
Good catch @tacerus! The version of QEMU we get with _debian:bullseye_ still creates the overlay and prints a warning, but the new one included with _debian:bookworm_ throws an error.

TL;DR: code needs to handle `vmdk` and `img` source files in addition to `qcow2`.

I checked the CI logs for vmx and noticed the bootstrap script just exits because the qemu process never starts:
```
2023-07-06 20:53:41,761: vrnetlab   DEBUG    Creating overlay disk image
2023-07-06 20:53:42,251: vrnetlab   DEBUG    Creating overlay disk image
qemu-img: /vmx/vfpc-3-overlay.img: Image is not in qcow2 format
Could not open backing image.
[ ... ]
2023-07-06 20:54:45,768: vrnetlab   INFO     Unable to connect to qemu monitor (port 4003), retrying in a second (attempt 58)
2023-07-06 20:54:46,769: vrnetlab   INFO     Unable to connect to qemu monitor (port 4003), retrying in a second (attempt 59)
2023-07-06 20:54:47,771: vrnetlab   INFO     Unable to connect to qemu monitor (port 4003), retrying in a second (attempt 60)
Traceback (most recent call last):
  File "/launch.py", line 385, in <module>
    vr.start()
  File "/launch.py", line 295, in start
    super(VMX, self).start()
  File "/vrnetlab.py", line 404, in start
    vm.work()
  File "/vrnetlab.py", line 301, in work
    self.check_qemu()
  File "/vrnetlab.py", line 318, in check_qemu
    self.start()
  File "/launch.py", line 243, in start
    super(VMX_vfpc, self).start()
  File "/vrnetlab.py", line 141, in start
    raise QemuBroken("Unable to connect to qemu monitor on port {}".format(4000 + self.num))
vrnetlab.QemuBroken: Unable to connect to qemu monitor on port 4003
```

This happens when we try to test the built container image for vmx. Building vmx is a 2-step process, first build the container image with all the tools and files. Then we run the container with `--install` to force the OS to "settle" itself once. The install phase only starts the control plane VMs with `qcow2` formatted images. But the final container image starts the forwarding plane VM too and that one uses `.img`. According to the [docs] this maps to `raw` format. AFAICT there are in total three source image extensions we use:
1. `qcow2`
2. `img` (`raw`?)
3. `vmdk`

[docs]: https://qemu-project.gitlab.io/qemu/system/images.html

**Comment by tacerus** [comment]
Hi @mzagozen, thank you for the detailed explanation. I hope my follow-up patch restores the compatibility with the other formats. Should more formats with obscure extensions be needed in the future, they can easily be added to the dict.
Could you run the CI again? Unfortunately I don't have images for the other formats handy to test it locally.

**Comment by tacerus** [comment]
Thanks! The vmx test seems to pass now, but there still is an issue with xrv9k. I can't notice any issues during startup in it, is it related to my change?

**Comment by mzagozen** [comment]
I have seen the xrv9k job fail before for unrelated changes. In the [job log] for the first run  see the test was successful for version 6.6.2 but not for 7.1.2. It passed in the second run. That proves your change is not responsible for that failure.

[job log]: https://github.com/vrnetlab/vrnetlab/actions/runs/5505637557/jobs/10048114976

**Comment by mzagozen** [comment]
I was wrong about my initial assumption that file extensions are a reliable source of truth. On a hunch I pushed the change through our internal CI where we have some extra (not yet upstreamed) platforms and in one case the extension did not match the actual format. This throws an error in `qemu-img create`. But I did find out that we can use `qemu-img info` to detect the format on the fly. It seems reasonably fast. Please take a look @ https://github.com/vrnetlab/vrnetlab/commit/9d08a0c27b56d0adfae5aa5b8dafad4c0539bcae

**Comment by tacerus** [comment]
That's unfortunate - and a bit ironic that we need to use `qemu-img` just to read information to pass back to a second `qemu-img` operation. ;-) I think your approach is good and I added a comment to the commit.

**Comment by tacerus** [comment]
Closing in favor of #326.
